File: python/ter/secrets/backends/vault.py
"""
HashiCorp Vault Secret Backend (Python Implementation)

Integrates with HashiCorp Vault for secure secret storage.
Supports token, AppRole, JWT, and Kubernetes authentication.
"""

from typing import Dict, List, Optional, Any, Union, Callable
from dataclasses import dataclass
import asyncio
import json
from abc import ABC

from ..secret_manager import SecretBackend, AuditFilter, AuditEntry


@dataclass
class VaultConfig:
    """Vault configuration"""
    address: str  # e.g., https://vault.example.com:8200
    secret_path: str = 'secret'  # e.g., secret/data
    auth: Dict[str, Any] = None  # Auth config (see below)
    tls_skip_verify: bool = False
    timeout: int = 5  # seconds


class VaultBackend(SecretBackend):
    """
    Vault Secret Backend Implementation

    Supports authentication methods:
    - Token: Direct token auth
    - AppRole: Role ID + Secret ID
    - JWT: JWT token with role
    - Kubernetes: K8s service account token
    """

    def __init__(self, config: VaultConfig):
        """Initialize Vault backend"""
        self.config = config
        self.config.address = self.config.address.rstrip('/')
        self.token: Optional[str] = None
        self.token_expiry: float = 0

        if not self.config.address:
            raise ValueError('Vault address is required')

        if not self.config.auth:
            raise ValueError('Vault auth configuration is required')

    @property
    def name(self) -> str:
        return 'vault'

    async def is_available(self) -> bool:
        """Check if Vault is available and authenticated"""
        try:
            # If we have a valid token, we're good
            import time
            if self.token and time.time() < self.token_expiry:
                return True

            # Try to authenticate
            await self._authenticate()
            return True
        except Exception:
            return False

    async def get_secret(self, key: str) -> Optional[str]:
        """Get a secret from Vault"""
        await self._ensure_authenticated()

        url = f"{self.config.address}/v1/{self.config.secret_path}/data/{key}"
        response = await self._request('GET', url)

        if response['status'] == 404:
            return None

        if response['status'] != 200:
            raise RuntimeError(
                f"Failed to get secret from Vault: {response['status']} {response.get('message')}"
            )

        data = json.loads(response.get('data', '{}'))
        # KV v2 API returns data.data.{key}
        return data.get('data', {}).get('data', {}).get(key)

    async def put_secret(self, key: str, value: str) -> None:
        """Put a secret in Vault"""
        await self._ensure_authenticated()

        url = f"{self.config.address}/v1/{self.config.secret_path}/data/{key}"
        body = json.dumps({'data': {key: value}})

        response = await self._request('POST', url, body)

        if response['status'] != 200:
            raise RuntimeError(
                f"Failed to put secret in Vault: {response['status']} {response.get('message')}"
            )

    async def delete_secret(self, key: str) -> None:
        """Delete a secret from Vault"""
        await self._ensure_authenticated()

        url = f"{self.config.address}/v1/{self.config.secret_path}/data/{key}"
        response = await self._request('DELETE', url)

        if response['status'] != 204:
            raise RuntimeError(
                f"Failed to delete secret from Vault: {response['status']} {response.get('message')}"
            )

    async def list_secrets(self, pattern: Optional[str] = None) -> List[str]:
        """List secrets in Vault"""
        await self._ensure_authenticated()

        if pattern:
            base_path = f"{self.config.secret_path}/data/{pattern}"
        else:
            base_path = f"{self.config.secret_path}/data"

        url = f"{self.config.address}/v1/{base_path}"
        response = await self._request('LIST', url)

        if response['status'] == 404:
            return []

        if response['status'] != 200:
            raise RuntimeError(
                f"Failed to list secrets from Vault: {response['status']} {response.get('message')}"
            )

        data = json.loads(response.get('data', '{}'))
        return data.get('data', {}).get('keys', [])

    async def rotate_secret(self, key: str) -> str:
        """Rotate a secret in Vault"""
        await self._ensure_authenticated()

        # Generate new value
        import secrets
        new_value = secrets.token_hex(32)

        # Update secret
        await self.put_secret(key, new_value)
        return new_value

    async def get_audit_log(self, filter: Optional[AuditFilter] = None) -> List[AuditEntry]:
        """Get audit log from Vault"""
        # Vault audit logs require special endpoint and permissions
        # Returning empty list as placeholder
        return []

    async def close(self) -> None:
        """Close connection"""
        self.token = None

    # Private helpers

    async def _ensure_authenticated(self) -> None:
        """Ensure we have a valid token"""
        import time
        if self.token and time.time() < self.token_expiry:
            return

        await self._authenticate()

    async def _authenticate(self) -> None:
        """Authenticate to Vault using configured method"""
        auth_config = self.config.auth
        method = auth_config.get('method')

        if method == 'token':
            self.token = auth_config.get('token')
            import time
            self.token_expiry = time.time() + 24 * 60 * 60  # 24 hours

        elif method == 'approle':
            await self._authenticate_approle(auth_config)

        elif method == 'jwt':
            await self._authenticate_jwt(auth_config)

        elif method == 'kubernetes':
            await self._authenticate_kubernetes(auth_config)

        else:
            raise ValueError(f"Unknown Vault auth method: {method}")

    async def _authenticate_approle(self, auth: Dict[str, Any]) -> None:
        """Authenticate using AppRole"""
        url = f"{self.config.address}/v1/auth/approle/login"
        body = json.dumps({
            'role_id': auth.get('role_id'),
            'secret_id': auth.get('secret_id'),
        })

        response = await self._request('POST', url, body, skip_auth=True)

        if response['status'] != 200:
            raise RuntimeError(
                f"AppRole authentication failed: {response['status']} {response.get('message')}"
            )

        data = json.loads(response.get('data', '{}'))
        self.token = data.get('auth', {}).get('client_token')
        lease_duration = data.get('auth', {}).get('lease_duration', 3600)

        import time
        self.token_expiry = time.time() + lease_duration

        if not self.token:
            raise RuntimeError('No token received from Vault')

    async def _authenticate_jwt(self, auth: Dict[str, Any]) -> None:
        """Authenticate using JWT"""
        jwt_token = auth.get('jwt')

        # If JWT is callable, call it to get the token
        if callable(jwt_token):
            jwt_token = await jwt_token() if asyncio.iscoroutinefunction(jwt_token) else jwt_token()

        url = f"{self.config.address}/v1/auth/jwt/login"
        body = json.dumps({
            'role': auth.get('role'),
            'jwt': jwt_token,
        })

        response = await self._request('POST', url, body, skip_auth=True)

        if response['status'] != 200:
            raise RuntimeError(
                f"JWT authentication failed: {response['status']} {response.get('message')}"
            )

        data = json.loads(response.get('data', '{}'))
        self.token = data.get('auth', {}).get('client_token')
        lease_duration = data.get('auth', {}).get('lease_duration', 3600)

        import time
        self.token_expiry = time.time() + lease_duration

        if not self.token:
            raise RuntimeError('No token received from Vault')

    async def _authenticate_kubernetes(self, auth: Dict[str, Any]) -> None:
        """Authenticate using Kubernetes service account"""
        # Read service account token from Kubernetes
        service_account_path = auth.get('service_account_path', '/var/run/secrets/kubernetes.io/serviceaccount')
        
        try:
            with open(f"{service_account_path}/token", 'r') as f:
                jwt_token = f.read().strip()
        except IOError as e:
            raise RuntimeError(f"Failed to read Kubernetes service account token: {e}")

        url = f"{self.config.address}/v1/auth/kubernetes/login"
        body = json.dumps({
            'role': auth.get('role'),
            'jwt': jwt_token,
        })

        response = await self._request('POST', url, body, skip_auth=True)

        if response['status'] != 200:
            raise RuntimeError(
                f"Kubernetes authentication failed: {response['status']} {response.get('message')}"
            )

        data = json.loads(response.get('data', '{}'))
        self.token = data.get('auth', {}).get('client_token')
        lease_duration = data.get('auth', {}).get('lease_duration', 3600)

        import time
        self.token_expiry = time.time() + lease_duration

        if not self.token:
            raise RuntimeError('No token received from Vault')

    async def _request(
        self,
        method: str,
        url: str,
        body: Optional[str] = None,
        skip_auth: bool = False,
    ) -> Dict[str, Any]:
        """Make HTTP request to Vault"""
        import aiohttp

        headers: Dict[str, str] = {
            'X-Vault-Request': 'true',
            'Content-Type': 'application/json',
        }

        if self.token and not skip_auth:
            headers['X-Vault-Token'] = self.token

        try:
            async with aiohttp.ClientSession() as session:
                async with session.request(
                    method,
                    url,
                    data=body,
                    headers=headers,
                    timeout=aiohttp.ClientTimeout(total=self.config.timeout),
                    ssl=not self.config.tls_skip_verify,
                ) as response:
                    data = await response.text()
                    return {
                        'status': response.status,
                        'message': response.reason,
                        'data': data,
                    }
        except Exception as e:
            raise RuntimeError(f"Vault request failed: {e}")

File: src/secrets/backends/vault.ts
/**
 * HashiCorp Vault Secret Backend
 *
 * Integrates with HashiCorp Vault for secure secret storage.
 * Supports multiple auth methods: token, AppRole, JWT, Kubernetes.
 */

import * as https from 'https';
import { SecretBackend, AuditFilter, AuditEntry } from '../secret-manager';

export interface VaultConfig {
  address: string; // e.g., https://vault.example.com:8200
  secretPath?: string; // e.g., secret/data (default: secret)
  auth: VaultAuthConfig;
  tlsSkipVerify?: boolean;
  timeout?: number;
}

export type VaultAuthConfig =
  | VaultTokenAuth
  | VaultAppRoleAuth
  | VaultJWTAuth
  | VaultKubernetesAuth;

export interface VaultTokenAuth {
  method: 'token';
  token: string;
}

export interface VaultAppRoleAuth {
  method: 'approle';
  roleId: string;
  secretId: string;
}

export interface VaultJWTAuth {
  method: 'jwt';
  role: string;
  jwt: string | (() => Promise<string>); // JWT token or callable to fetch it
}

export interface VaultKubernetesAuth {
  method: 'kubernetes';
  role: string;
  serviceAccountPath?: string;
}

/**
 * Vault Backend Implementation
 */
export class VaultBackend implements SecretBackend {
  name = 'vault';
  private config: VaultConfig;
  private token: string | null = null;
  private tokenExpiry: number = 0;
  private requestTimeout: NodeJS.Timeout | null = null;

  constructor(config: VaultConfig) {
    this.config = {
      secretPath: 'secret',
      ...config,
    };

    if (!this.config.address) {
      throw new Error('Vault address is required');
    }

    // Normalize address (remove trailing slash)
    this.config.address = this.config.address.replace(/\/$/, '');
  }

  /**
   * Check if Vault is available and authenticated
   */
  async isAvailable(): Promise<boolean> {
    try {
      // If we have a valid token, we're good
      if (this.token && Date.now() < this.tokenExpiry) {
        return true;
      }

      // Try to authenticate
      await this.authenticate();
      return true;
    } catch {
      return false;
    }
  }

  /**
   * Get a secret from Vault
   */
  async getSecret(key: string): Promise<string | null> {
    await this.ensureAuthenticated();

    const response = await this.request('GET', `/v1/${this.config.secretPath}/data/${key}`);

    if (response.statusCode === 404) {
      return null;
    }

    if (response.statusCode !== 200) {
      throw new Error(
        `Failed to get secret from Vault: ${response.statusCode} ${response.statusMessage}`
      );
    }

    const data = JSON.parse(response.data);
    return data.data?.data?.[key] || data.data?.data?.value || null;
  }

  /**
   * Put a secret in Vault
   */
  async putSecret(key: string, value: string): Promise<void> {
    await this.ensureAuthenticated();

    const body = JSON.stringify({
      data: {
        [key]: value,
      },
    });

    const response = await this.request(
      'POST',
      `/v1/${this.config.secretPath}/data/${key}`,
      body
    );

    if (response.statusCode !== 200) {
      throw new Error(
        `Failed to put secret in Vault: ${response.statusCode} ${response.statusMessage}`
      );
    }
  }

  /**
   * Delete a secret from Vault
   */
  async deleteSecret(key: string): Promise<void> {
    await this.ensureAuthenticated();

    const response = await this.request('DELETE', `/v1/${this.config.secretPath}/data/${key}`);

    if (response.statusCode !== 204) {
      throw new Error(
        `Failed to delete secret from Vault: ${response.statusCode} ${response.statusMessage}`
      );
    }
  }

  /**
   * List secrets in Vault
   */
  async listSecrets(pattern?: string): Promise<string[]> {
    await this.ensureAuthenticated();

    const basePath = pattern
      ? `${this.config.secretPath}/data/${pattern}`
      : `${this.config.secretPath}/data`;

    const response = await this.request('LIST', `/v1/${basePath}`);

    if (response.statusCode === 404) {
      return [];
    }

    if (response.statusCode !== 200) {
      throw new Error(
        `Failed to list secrets from Vault: ${response.statusCode} ${response.statusMessage}`
      );
    }

    const data = JSON.parse(response.data);
    return data.data?.keys || [];
  }

  /**
   * Rotate a secret in Vault (database specific)
   */
  async rotateSecret(key: string): Promise<string> {
    await this.ensureAuthenticated();

    // Rotate by setting a new value
    const newValue = this.generateRandomSecret();
    await this.putSecret(key, newValue);
    return newValue;
  }

  /**
   * Get audit log from Vault
   */
  async getAuditLog(filter?: AuditFilter): Promise<AuditEntry[]> {
    // Vault audit logs require special endpoint and permissions
    // This is a simplified implementation
    const entries: AuditEntry[] = [];

    // In a real implementation, you'd query Vault's audit backend
    // For now, return empty to indicate feature is available but not fully implemented
    return entries;
  }

  /**
   * Close connection
   */
  async close(): Promise<void> {
    if (this.requestTimeout) {
      clearTimeout(this.requestTimeout);
    }
  }

  // Private helpers

  private async ensureAuthenticated(): Promise<void> {
    // If token is still valid, skip re-authentication
    if (this.token && Date.now() < this.tokenExpiry) {
      return;
    }

    await this.authenticate();
  }

  private async authenticate(): Promise<void> {
    const authConfig = this.config.auth;

    switch (authConfig.method) {
      case 'token':
        this.token = authConfig.token;
        this.tokenExpiry = Date.now() + 24 * 60 * 60 * 1000; // 24 hours
        break;

      case 'approle':
        await this.authenticateAppRole(authConfig);
        break;

      case 'jwt':
        await this.authenticateJWT(authConfig);
        break;

      case 'kubernetes':
        await this.authenticateKubernetes(authConfig);
        break;

      default:
        throw new Error(`Unknown auth method: ${(authConfig as any).method}`);
    }
  }

  private async authenticateAppRole(auth: VaultAppRoleAuth): Promise<void> {
    const body = JSON.stringify({
      role_id: auth.roleId,
      secret_id: auth.secretId,
    });

    const response = await this.request('POST', '/v1/auth/approle/login', body, true);

    if (response.statusCode !== 200) {
      throw new Error(
        `AppRole authentication failed: ${response.statusCode} ${response.statusMessage}`
      );
    }

    const data = JSON.parse(response.data);
    this.token = data.auth?.client_token;
    this.tokenExpiry = Date.now() + (data.auth?.lease_duration || 3600) * 1000;

    if (!this.token) {
      throw new Error('No token received from Vault');
    }
  }

  private async authenticateJWT(auth: VaultJWTAuth): Promise<void> {
    const jwt = typeof auth.jwt === 'function' ? await auth.jwt() : auth.jwt;

    const body = JSON.stringify({
      role: auth.role,
      jwt,
    });

    const response = await this.request('POST', '/v1/auth/jwt/login', body, true);

    if (response.statusCode !== 200) {
      throw new Error(
        `JWT authentication failed: ${response.statusCode} ${response.statusMessage}`
      );
    }

    const data = JSON.parse(response.data);
    this.token = data.auth?.client_token;
    this.tokenExpiry = Date.now() + (data.auth?.lease_duration || 3600) * 1000;

    if (!this.token) {
      throw new Error('No token received from Vault');
    }
  }

  private async authenticateKubernetes(auth: VaultKubernetesAuth): Promise<void> {
    // Read service account token from Kubernetes
    const fs = require('fs').promises;
    const serviceAccountPath = auth.serviceAccountPath || '/var/run/secrets/kubernetes.io/serviceaccount';
    const token = await fs.readFile(`${serviceAccountPath}/token`, 'utf-8');

    const body = JSON.stringify({
      role: auth.role,
      jwt: token.trim(),
    });

    const response = await this.request('POST', '/v1/auth/kubernetes/login', body, true);

    if (response.statusCode !== 200) {
      throw new Error(
        `Kubernetes authentication failed: ${response.statusCode} ${response.statusMessage}`
      );
    }

    const data = JSON.parse(response.data);
    this.token = data.auth?.client_token;
    this.tokenExpiry = Date.now() + (data.auth?.lease_duration || 3600) * 1000;

    if (!this.token) {
      throw new Error('No token received from Vault');
    }
  }

  private async request(
    method: string,
    path: string,
    body?: string,
    skipAuth?: boolean
  ): Promise<VaultResponse> {
    return new Promise((resolve, reject) => {
      const url = new URL(this.config.address + path);

      const options = {
        hostname: url.hostname,
        port: url.port || 8200,
        path: url.pathname + url.search,
        method,
        headers: {
          'Content-Type': 'application/json',
          'X-Vault-Request': 'true',
        } as Record<string, string>,
        rejectUnauthorized: !this.config.tlsSkipVerify,
        timeout: this.config.timeout || 5000,
      };

      if (this.token && !skipAuth) {
        options.headers['X-Vault-Token'] = this.token;
      }

      if (body) {
        options.headers['Content-Length'] = String(Buffer.byteLength(body));
      }

      const req = https.request(options, (res) => {
        let data = '';

        res.on('data', (chunk) => {
          data += chunk;
        });

        res.on('end', () => {
          resolve({
            statusCode: res.statusCode || 500,
            statusMessage: res.statusMessage,
            data,
            headers: res.headers,
          });
        });
      });

      req.on('error', reject);

      if (body) {
        req.write(body);
      }

      req.end();
    });
  }

  private generateRandomSecret(): string {
    return require('crypto').randomBytes(32).toString('hex');
  }
}

interface VaultResponse {
  statusCode: number;
  statusMessage?: string;
  data: string;
  headers: Record<string, any>;
}

export default VaultBackend;
